### ISSUE TITLE
Trading/Liquid Hours Bad Data

### ISSUE THREAD

**Issue by ejackson** [cbilson/ib-re-actor#8]
I'm busy creating some proposal code for parsing trading hours.  The IB trading hours reported through request-contract-details for ESH3 today, Jan 14 2013 is

"20130114:1700-1515,1530-1615;20130115:1700-1515,1530-1615"

The problem is the first interval.   Clearly they haven't been trading since Friday at 17:00, nor Sunday at 17:00.  So what does this mean ?  How to return it ?

Liquid hours are sensibly reported as 

"20130114:0830-1515;20130115:0830-1515"

I'm going to trawl the mailing list to see what I can gather, but if anybody knows more on the topic please tell me !

**Comment by cbilson** [comment]
They actually do trade on Sunday. Here is the CME schedule, for example:
http://www.cmegroup.com/trading_hours/.

This might be kind of tricky, but I think maybe you could assume that if
the open time was later in the day than the close time, it must open on the
previous day. For holidays like Christmas, that can occur in the middle of
the week, they usually have a modified open the next day (Dec. 26th) that
is very early, thus before the close time.

I bet there are other edge cases we are not thinking of, but I think that's
par for the course. Trading hours for global markets are actually a lot
more volatile than people think, with lots of last minute changes or
unexpected closures due to accidents or disasters. They usually have
discretion to modify the hours in real time if they need to as well, so you
never really know. So I guess there is no way to get this perfect, but just
being able to use the data from IB will be a huge help.

**Comment by ejackson** [comment]
OK cool, I'll get this done and commit for you to review.

**Comment by ejackson** [comment]
Committed, but forgot to link the issue.  Getting late !

https://github.com/cbilson/ib-re-actor/commit/c7f4b58125e57f6ac7d38bbc944a7ae3b64cb727

A couple of points
1. IB reports the trading hours in their local time.  I've moved all of that to UTC.
2. Whenever IB reports a date/time as EST I've discovered (through suffering) that they mean America/New_York so have amended the code to address this problem.
3. I've signalled a closed day with a zero length interval so that it can be distinguished from simply missing data.  Is this agreeable ?

The code is pretty scratchy for now, but functions.  

Comments ?
